namer/consul/src/main/scala/io/buoyant/namer/consul/DcServices.scala

```diff
@@ -0,0 +1,96 @@
+package io.buoyant.namer.consul
+
+import com.twitter.finagle.http.{Response, Status}
+import com.twitter.finagle.{Addr, Failure}
+import com.twitter.util._
+import io.buoyant.consul.v1
+
+private[consul] object DcServices {
+
+  /**
+   * Contains all cached serviceMap responses and the mapping of names
+   * to Addrs for a particular datacenter.
+   */
+  def apply(
+    consulApi: v1.ConsulApi,
+    name: String,
+    domain: Option[String]
+  ): Activity[Map[SvcKey, Var[Addr]]] = {
+
+    def getServices(index: Option[String]): Future[v1.Indexed[Set[SvcKey]]] =
+      consulApi.serviceMap(
+        datacenter = Some(name),
+        blockingIndex = index,
+        retry = true
+      ).map(toServices)
+
+    val states = Var.async[Activity.State[Map[SvcKey, Var[Addr]]]](Activity.Pending) { state =>
+      @volatile var stopped: Boolean = false
+
+      def loop(index0: Option[String], cache: Map[SvcKey, Var[Addr]]): Future[Unit] =
+        if (stopped) Future.Unit
+        else getServices(index0).transform {
+          case Throw(e@v1.UnexpectedResponse(rsp)) if rsp.status == Status.NotFound =>
```

Comment:
Also, it's wrong to expect `404` in such situation (as per comment below) because:
```
$ curl -i "http://localhost:8500/v1/catalog/services?dc=WRONG"
HTTP/1.1 500 Internal Server Error
X-Consul-Index: 0
X-Consul-Knownleader: false
X-Consul-Lastcontact: 0
Date: Wed, 07 Sep 2016 20:09:43 GMT
Content-Length: 32
Content-Type: text/plain; charset=utf-8

rpc error: No path to datacenter
```